src/popylar_prf/stimulus.py

```diff
@@ -0,0 +1,248 @@
+    >>> design = np.ones((num_frames, width, height))
+    >>> pixel_size = 0.05
+    >>> x = (np.arange(width) - (width - 1) / 2) * pixel_size
+    >>> y = (np.arange(height) - (height - 1) / 2) * pixel_size
+    >>> xv, yv = np.meshgrid(x, y)
+    >>> grid = np.stack((xv, yv), axis=-1)  # shape (height, width, 2)
+    >>> grid = np.stack((xv, yv), axis=-1)  # shape (height, width, 2)
+    >>> # The coordinates of the bottom-left corner:
+    >>> grid[0, 0, :]
+    array([-0.375, -0.375])
+    >>> # The coordinates of the top-right corner:
+    >>> grid[15, 15, :]
+    array([0.375, 0.375])
+    >>> Stimulus(design=design, grid=grid, dimension_labels=["y", "x"])
+    Stimulus(design=array[10, 16, 16], grid=array[16, 16, 2], dimension_labels=['y', 'x'])
+
+    """
+
+    __hash__ = None  # We don't want the object to be hashable because it's mutable
+
+    def __init__(self, design: np.ndarray, grid: np.ndarray, dimension_labels: Sequence[str] | None = None):
+        self.design = design
+        self.grid = grid
+        self.dimension_labels = dimension_labels
+
+        self._check_grid_design_shape()
+        self._check_grid_dimensions()
+        self._check_dimension_labels()
+
+    def _check_grid_design_shape(self) -> None:
+        if not self.design.shape[1:] == self.grid.shape[:-1]:
+            raise GridDesignShapeError(self.design.shape, self.grid.shape)
+
+    def _check_grid_dimensions(self) -> None:
+        if not len(self.grid.shape[:-1]) == self.grid.shape[-1]:
+            raise GridDimensionsError(self.grid.shape)
+
+    def _check_dimension_labels(self) -> None:
+        if self.dimension_labels is not None and not self.grid.shape[-1] == len(self.dimension_labels):
+            raise DimensionLabelsError(len(self.dimension_labels), self.grid.shape[-1])
+
+    def __repr__(self) -> str:
+        return f"""Stimulus(
+            design={np.array_repr(self.design)},
+            grid={np.array_repr(self.grid)},
+            dimension_labels={self.dimension_labels}
+        )"""
+
+    def __str__(self) -> str:
+        design_shape_str = ", ".join([str(s) for s in self.design.shape])
+        grid_shape_str = ", ".join([str(s) for s in self.grid.shape])
+
+        return f"Stimulus(design=array[{design_shape_str}], grid=array[{grid_shape_str}], dimension_labels={self.dimension_labels})"  # noqa: E501
+
+    def __eq__(self, value: object) -> bool:
+        if not isinstance(value, Stimulus):
+            msg = "Stimulus objects can only be compared against other Stimulus objects"
+            raise TypeError(msg)
+
+        if self.design.shape != value.design.shape or self.grid.shape != value.grid.shape:
+            return False
+
+        design_equal = np.all(self.design == value.design)
+        grid_equal = np.all(self.grid == value.grid)
+        dimensions_labels_equal = self.dimension_labels == value.dimension_labels
+
+        return bool(design_equal and grid_equal and dimensions_labels_equal)
```

Comment:
```suggestion
    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Stimulus):
            msg = "Stimulus objects can only be compared against other Stimulus objects"
            raise TypeError(msg)

        if self.design.shape != other.design.shape or self.grid.shape != other.grid.shape:
            return False

        design_equal = np.all(self.design == other.design)
        grid_equal = np.all(self.grid == other.grid)
        dimensions_labels_equal = self.dimension_labels == other.dimension_labels

        return bool(design_equal and grid_equal and dimensions_labels_equal)
```
I'd find this more readable but am fine with either option.